In [30]:
import numpy as np
import pandas as pd
from pathlib import Path

In [31]:
train_df = pd.read_csv(Path('Resources/2019loans.csv'))
test_df = pd.read_csv(Path('Resources/2020Q1loans.csv'))
train_df

Unnamed: 0   index  loan_amnt  int_rate  installment home_ownership  \
0           57107   57107    13375.0    0.1797       483.34       MORTGAGE   
1          141451  141451    21000.0    0.1308       478.68       MORTGAGE   
2          321143  321143    20000.0    0.1240       448.95       MORTGAGE   
3           11778   11778     3000.0    0.1240       100.22           RENT   
4          169382  169382    30000.0    0.1612      1056.49       MORTGAGE   
...           ...     ...        ...       ...          ...            ...   
12175      354912  354912    19975.0    0.2565       801.09           RENT   
12176      354944  354944    15000.0    0.1774       540.34           RENT   
12177      354973  354973     3600.0    0.1862       131.28           RENT   
12178      355002  355002    15000.0    0.0881       475.68       MORTGAGE   
12179      355307  355307    15000.0    0.1774       540.34           RENT   

       annual_inc verification_status loan_status pymnt_plan  ...  \
0        223000.0        Not Verified    low_risk          n  ...   
1        123000.0     Source Verified    low_risk          n  ...   
2        197000.0     Source Verified    low_risk          n  ...   
3         45000.0        Not Verified    low_risk          n  ...   
4        133000.0     Source Verified    low_risk          n  ...   
...           ...                 ...         ...        ...  ...   
12175     28000.0        Not Verified   high_risk          n  ...   
12176     50000.0            Verified   high_risk          n  ...   
12177     60000.0        Not Verified   high_risk          n  ...   
12178     62000.0     Source Verified   high_risk          n  ...   
12179     54080.0        Not Verified   high_risk          n  ...   

       pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
0               100.0              50.0                   0.0        0.0   
1                85.0              33.3                   0.0        0.0   
2                85.7              33.3                   0.0        0.0   
3               100.0              16.7                   1.0        0.0   
4               100.0              66.7                   0.0        0.0   
...               ...               ...                   ...        ...   
12175           100.0              16.7                   0.0        0.0   
12176            90.5              11.1                   0.0        0.0   
12177           100.0               0.0                   0.0        0.0   
12178           100.0               0.0                   0.0        0.0   
12179            90.0             100.0                   0.0        0.0   

       tot_hi_cred_lim  total_bal_ex_mort  total_bc_limit  \
0             577150.0           122018.0         32000.0   
1             132750.0            27896.0         15900.0   
2             628160.0           114043.0         22600.0   
3              42006.0            20761.0         19900.0   
4             283248.0           109056.0         79500.0   
...                ...                ...             ...   
12175          50055.0            28192.0         18700.0   
12176          70324.0            57025.0         13300.0   
12177          83765.0            55156.0         14800.0   
12178         189930.0            23748.0          7000.0   
12179          48968.0            29786.0          2900.0   

      total_il_high_credit_limit  hardship_flag  debt_settlement_flag  
0                       170200.0              N                     N  
1                        35398.0              N                     N  
2                        90340.0              N                     N  
3                        15406.0              N                     N  
4                        58778.0              N                     N  
...                          ...            ...                   ...  
12175                    19055.0              N                     N  
12176              

In [32]:
test_df

Unnamed: 0  index  loan_amnt  int_rate  installment home_ownership  \
0          67991  67991    40000.0    0.0819       814.70       MORTGAGE   
1          25429  25429     6000.0    0.1524       208.70           RENT   
2          38496  38496     3600.0    0.1695       128.27           RENT   
3          19667  19667    20000.0    0.1524       478.33           RENT   
4          37505  37505     3600.0    0.1240       120.27           RENT   
...          ...    ...        ...       ...          ...            ...   
4697       77282  77282    30000.0    0.1240       673.42           RENT   
4698       77291  77291    24000.0    0.0756       747.22           RENT   
4699       77292  77292    10000.0    0.2305       387.36           RENT   
4700       77297  77297     8000.0    0.1862       205.86           RENT   
4701       77304  77304    30000.0    0.2055      1123.34           RENT   

      annual_inc verification_status loan_status pymnt_plan  ...  \
0       140000.0        Not Verified    low_risk          n  ...   
1        55000.0        Not Verified    low_risk          n  ...   
2        42000.0        Not Verified    low_risk          n  ...   
3       100000.0        Not Verified    low_risk          n  ...   
4        50000.0        Not Verified    low_risk          n  ...   
...          ...                 ...         ...        ...  ...   
4697    140480.0     Source Verified   high_risk          n  ...   
4698     50000.0        Not Verified   high_risk          n  ...   
4699     33000.0            Verified   high_risk          n  ...   
4700     38000.0     Source Verified   high_risk          n  ...   
4701    180000.0     Source Verified   high_risk          n  ...   

      pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
0               97.7               0.0                   0.0        0.0   
1               66.7               0.0                   0.0        0.0   
2              100.0               0.0                   0.0        0.0   
3              100.0              50.0                   0.0        0.0   
4              100.0              25.0                   0.0        0.0   
...              ...               ...                   ...        ...   
4697           100.0              28.6                   0.0        0.0   
4698           100.0               0.0                   0.0        0.0   
4699           100.0               0.0                   0.0        0.0   
4700            95.0               0.0                   1.0        0.0   
4701           100.0              33.3                   0.0        0.0   

      tot_hi_cred_lim  total_bal_ex_mort  total_bc_limit  \
0            527975.0            70914.0         74600.0   
1             34628.0            23460.0          5900.0   
2             23100.0            19183.0          7300.0   
3             56481.0            43817.0         13800.0   
4             45977.0            32448.0         21000.0   
...               ...                ...             ...   
4697         159688.0           110873.0         48400.0   
4698          62375.0            18928.0         13300.0   
4699          43250.0            33022.0          8500.0   
4700          31357.0            19595.0          1500.0   
4701         218686.0           209389.0          5600.0   

     total_il_high_credit_limit  hardship_flag  debt_settlement_flag  
0                       99475.0              N                     N  
1                       23628.0              N                     N  
2                       15000.0              N                     N  
3                       35981.0              N                     N  
4                       24977.0              N                     N  
...                         ...            ...                   ...  
4697                   107388.0              N                     N  
4698                    30775.0              N                     N  
4699           

# Drop the null columns where all values are null
# Drop the null rows

In [33]:
train_df = train_df.dropna(axis='columns', how='all')
train_df = train_df.dropna()
test_df = test_df.dropna(axis='columns', how='all')
test_df = test_df.dropna()

# Convert target column "loan_status" into numeric

In [34]:
risk={"low_risk":0, "high_risk":1}
train_df = train_df.replace({'loan_status': risk})
test_df = test_df.replace({'loan_status': risk})

In [35]:
#Check our training data columns
train_df.columns

Index(['Unnamed: 0', 'index', 'loan_amnt', 'int_rate', 'installment',
       'home_ownership', 'annual_inc', 'verification_status', 'loan_status',
       'pymnt_plan', 'dti', 'delinq_2yrs', 'inq_last_6mths', 'open_acc',
       'pub_rec', 'revol_bal', 'total_acc', 'initial_list_status', 'out_prncp',
       'out_prncp_inv', 'total_pymnt', 'total_pymnt_inv', 'total_rec_prncp',
       'total_rec_int', 'total_rec_late_fee', 'recoveries',
       'collection_recovery_fee', 'last_pymnt_amnt',
       'collections_12_mths_ex_med', 'policy_code', 'application_type',
       'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal', 'open_acc_6m',
       'open_act_il', 'open_il_12m', 'open_il_24m', 'mths_since_rcnt_il',
       'total_bal_il', 'il_util', 'open_rv_12m', 'open_rv_24m', 'max_bal_bc',
       'all_util', 'total_rev_hi_lim', 'inq_fi', 'total_cu_tl', 'inq_last_12m',
       'acc_open_past_24mths', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util',
       'chargeoff_within_12_mths', 'delinq_amnt', 'mo_sin_ol

# Create our features
## One-hot-encoding to transform our feature into numeric

In [36]:
# Convert categorical data to numeric and separate target feature for training data
# Create our features
# Convert string columns to numbers and drop 'loan_status' column then assign to X
X_train = train_df.drop(['loan_status'], axis=1)

##One-hot-encoding to transform our feature into numeric
X_train = pd.get_dummies(X_train)


# Create our target
y_train = train_df['loan_status']
X_train.head()

Unnamed: 0   index  loan_amnt  int_rate  installment  annual_inc    dti  \
0       57107   57107    13375.0    0.1797       483.34    223000.0  29.99   
1      141451  141451    21000.0    0.1308       478.68    123000.0  11.26   
2      321143  321143    20000.0    0.1240       448.95    197000.0  11.28   
3       11778   11778     3000.0    0.1240       100.22     45000.0  18.08   
4      169382  169382    30000.0    0.1612      1056.49    133000.0  27.77   

   delinq_2yrs  inq_last_6mths  open_acc  ...  verification_status_Verified  \
0          0.0             0.0      15.0  ...                             0   
1          2.0             0.0      16.0  ...                             0   
2          0.0             0.0      12.0  ...                             0   
3          0.0             0.0      12.0  ...                             0   
4          0.0             2.0      13.0  ...                             0   

   pymnt_plan_n  initial_list_status_f  initial_list_status_w  \
0             1                      0                      1   
1             1                      0                      1   
2             1                      0                      1   
3             1                      0                      1   
4             1                      0                      1   

   application_type_Individual  application_type_Joint App  hardship_flag_N  \
0                            1                           0                1   
1                            1                           0                1   
2                            1                           0                1   
3                            1                           0                1   
4                            1                           0                1   

   hardship_flag_Y  debt_settlement_flag_N  debt_settlement_flag_Y  
0                0                       1                       0  
1                0                       1                       0  
2                0                       1                       0  
3                0                       1                       0  
4                0                       1                       0  

[5 rows x 94 columns]

In [37]:
y_train.value_counts()

0    6090
1    6090
Name: loan_status, dtype: int64

In [ ]:
test_df.columns

In [38]:
# Convert categorical data to numeric and separate target feature for testing data
# Convert categorical data to numeric and separate target feature for training data
# Create our features
# Convert string columns to numbers and drop 'loan_status' column then assign to X
X_test = test_df.drop(['loan_status'], axis=1)

##One-hot-encoding to transform our feature into numeric
X_test = pd.get_dummies(X_test)


# Create our target
y_test = test_df['loan_status']
X_test.head()

Unnamed: 0  index  loan_amnt  int_rate  installment  annual_inc    dti  \
0       67991  67991    40000.0    0.0819       814.70    140000.0  19.75   
1       25429  25429     6000.0    0.1524       208.70     55000.0  11.52   
2       38496  38496     3600.0    0.1695       128.27     42000.0   6.74   
3       19667  19667    20000.0    0.1524       478.33    100000.0  12.13   
4       37505  37505     3600.0    0.1240       120.27     50000.0  16.08   

   delinq_2yrs  inq_last_6mths  open_acc  ...  \
0          0.0             1.0      18.0  ...   
1          2.0             0.0       8.0  ...   
2          0.0             0.0       6.0  ...   
3          0.0             2.0       7.0  ...   
4          0.0             3.0       6.0  ...   

   verification_status_Source Verified  verification_status_Verified  \
0                                    0                             0   
1                                    0                             0   
2                                    0                             0   
3                                    0                             0   
4                                    0                             0   

   pymnt_plan_n  initial_list_status_f  initial_list_status_w  \
0             1                      0                      1   
1             1                      0                      1   
2             1                      0                      1   
3             1                      0                      1   
4             1                      0                      1   

   application_type_Individual  application_type_Joint App  hardship_flag_N  \
0                            1                           0                1   
1                            1                           0                1   
2                            1                           0                1   
3                            1                           0                1   
4                            1                           0                1   

   hardship_flag_Y  debt_settlement_flag_N  
0                0                       1  
1                0                       1  
2                0                       1  
3                0                       1  
4                0                       1  

[5 rows x 93 columns]

In [39]:
y_test.value_counts()

0    2351
1    2351
Name: loan_status, dtype: int64

In [40]:
# add missing dummy variables to testing set
# test_df(2020Q1loans) only have  93 column, use dummies function and add one column into our feature.
dummies=pd.get_dummies(test_df['debt_settlement_flag']).rename(columns={'N':"debt_settlement_flag_Y"})
X_test=pd.concat([X_test, dummies], axis=1)
X_test

Unnamed: 0  index  loan_amnt  int_rate  installment  annual_inc    dti  \
0          67991  67991    40000.0    0.0819       814.70    140000.0  19.75   
1          25429  25429     6000.0    0.1524       208.70     55000.0  11.52   
2          38496  38496     3600.0    0.1695       128.27     42000.0   6.74   
3          19667  19667    20000.0    0.1524       478.33    100000.0  12.13   
4          37505  37505     3600.0    0.1240       120.27     50000.0  16.08   
...          ...    ...        ...       ...          ...         ...    ...   
4697       77282  77282    30000.0    0.1240       673.42    140480.0  15.74   
4698       77291  77291    24000.0    0.0756       747.22     50000.0  26.81   
4699       77292  77292    10000.0    0.2305       387.36     33000.0  38.51   
4700       77297  77297     8000.0    0.1862       205.86     38000.0  16.36   
4701       77304  77304    30000.0    0.2055      1123.34    180000.0  12.06   

      delinq_2yrs  inq_last_6mths  open_acc  ...  \
0             0.0             1.0      18.0  ...   
1             2.0             0.0       8.0  ...   
2             0.0             0.0       6.0  ...   
3             0.0             2.0       7.0  ...   
4             0.0             3.0       6.0  ...   
...           ...             ...       ...  ...   
4697          0.0             0.0      20.0  ...   
4698          0.0             0.0       9.0  ...   
4699          0.0             2.0       7.0  ...   
4700          0.0             1.0       8.0  ...   
4701          0.0             0.0       8.0  ...   

      verification_status_Verified  pymnt_plan_n  initial_list_status_f  \
0                                0             1                      0   
1                                0             1                      0   
2                                0             1                      0   
3                                0             1                      0   
4                                0             1                      0   
...                            ...           ...                    ...   
4697                             0             1                      1   
4698                             0             1                      0   
4699                             1             1                      1   
4700                             0             1                      0   
4701                             0             1                      1   

      initial_list_status_w  application_type_Individual  \
0                         1                            1   
1                         1                            1   
2                         1                            1   
3                         1                            1   
4                         1                            1   
...                     ...                          ...   
4697                      0                            1   
4698                      1                            1   
4699                      0                            1   
4700                      1                            1   
4701                      0                            1   

      application_type_Joint App  hardship_flag_N  hardship_flag_Y  \
0                              0                1                0   
1                              0                1                0   
2                              0                1                0   
3                              0                1                0   
4                              0                1                0   
...                          ...              ...              ...   
4697                           0                1                0   
4698                           0                1                0   
4699                           0                1                0   
4700                           0                1                0   
4701                        

In [42]:
# Check balances
from collections import Counter
print(Counter(y_train))
print(Counter(y_test))

Counter({0: 6090, 1: 6090})
Counter({0: 2351, 1: 2351})


# Create Logistic Regression model

In [43]:
# Train the Logistic Regression model on the unscaled data and print the model score
#Create Logistic Regression model
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression()
classifier.fit(X_train, y_train)

/Users/chiachinghsieh/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:765: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


LogisticRegression()

In [44]:
#See the score of our Logistic Regression model
print(f"Training Data Score: {classifier.score(X_train, y_train)}")
print(f"Testing Data Score: {classifier.score(X_test, y_test)}")

Training Data Score: 0.648440065681445
Testing Data Score: 0.5253083794130158


In [45]:
#Use our Logistic Regression model to test the 2020Q1loans accuracy
predictions = classifier.predict(X_test)
print(f"The new point was classified as: {predictions}")

The new point was classified as: [0 0 0 ... 0 0 1]


In [46]:
pd.DataFrame({"Prediction": predictions, "Actual":y_test})

Prediction  Actual
0              0       0
1              0       0
2              0       0
3              1       0
4              0       0
...          ...     ...
4697           0       1
4698           0       1
4699           0       1
4700           0       1
4701           1       1

[4702 rows x 2 columns]

In [47]:
#print the compare reslut of prediction and actual 
predicted_labels = classifier.predict(X_test)
actual_labels = y_test
loan_risk = ['low_risk','high_risk']

def compare_labels(predicted_labels, actual_labels):
    """ a more neatly formatted comparison """
    NUM_LABELS = len(predicted_labels)
    num_correct = 0
    
    for i in range(NUM_LABELS):
        p = int(round(predicted_labels[i]))         # round protects from fp error 
        a = int(round(actual_labels[i]))
        result = "incorrect"
        if p == a:  # if they match,
            result = ""       # no longer incorrect
            num_correct += 1  # and we count a match!

        print(f"row {i:>3d} : {loan_risk[p]:>12s} {loan_risk[a]:<12s}   {result}")   

    print()
    print("Correct:", num_correct, "out of", NUM_LABELS)
    return num_correct

compare_labels(predicted_labels,actual_labels)

row   0 :     low_risk low_risk       
row   1 :     low_risk low_risk       
row   2 :     low_risk low_risk       
row   3 :    high_risk low_risk       incorrect
row   4 :     low_risk low_risk       
row   5 :     low_risk low_risk       
row   6 :     low_risk low_risk       
row   7 :     low_risk low_risk       
row   8 :     low_risk low_risk       
row   9 :     low_risk low_risk       
row  10 :     low_risk low_risk       
row  11 :     low_risk low_risk       
row  12 :     low_risk low_risk       
row  13 :     low_risk low_risk       
row  14 :    high_risk low_risk       incorrect
row  15 :     low_risk low_risk       
row  16 :     low_risk low_risk       
row  17 :     low_risk low_risk       
row  18 :     low_risk low_risk       
row  19 :     low_risk low_risk       
row  20 :     low_risk low_risk       
row  21 :     low_risk low_risk       
row  22 :     low_risk low_risk       
row  23 :     low_risk low_risk       
row  24 :     low_risk low_risk       
row  25

row 1135 :     low_risk low_risk       
row 1136 :     low_risk low_risk       
row 1137 :     low_risk low_risk       
row 1138 :     low_risk low_risk       
row 1139 :     low_risk low_risk       
row 1140 :     low_risk low_risk       
row 1141 :    high_risk low_risk       incorrect
row 1142 :     low_risk low_risk       
row 1143 :     low_risk low_risk       
row 1144 :     low_risk low_risk       
row 1145 :     low_risk low_risk       
row 1146 :     low_risk low_risk       
row 1147 :     low_risk low_risk       
row 1148 :     low_risk low_risk       
row 1149 :     low_risk low_risk       
row 1150 :     low_risk low_risk       
row 1151 :     low_risk low_risk       
row 1152 :     low_risk low_risk       
row 1153 :     low_risk low_risk       
row 1154 :     low_risk low_risk       
row 1155 :     low_risk low_risk       
row 1156 :    high_risk low_risk       incorrect
row 1157 :     low_risk low_risk       
row 1158 :     low_risk low_risk       
row 1159 :     low_ris

row 2135 :     low_risk low_risk       
row 2136 :     low_risk low_risk       
row 2137 :     low_risk low_risk       
row 2138 :     low_risk low_risk       
row 2139 :     low_risk low_risk       
row 2140 :     low_risk low_risk       
row 2141 :     low_risk low_risk       
row 2142 :     low_risk low_risk       
row 2143 :     low_risk low_risk       
row 2144 :     low_risk low_risk       
row 2145 :     low_risk low_risk       
row 2146 :     low_risk low_risk       
row 2147 :     low_risk low_risk       
row 2148 :     low_risk low_risk       
row 2149 :     low_risk low_risk       
row 2150 :     low_risk low_risk       
row 2151 :     low_risk low_risk       
row 2152 :     low_risk low_risk       
row 2153 :     low_risk low_risk       
row 2154 :     low_risk low_risk       
row 2155 :     low_risk low_risk       
row 2156 :     low_risk low_risk       
row 2157 :     low_risk low_risk       
row 2158 :     low_risk low_risk       
row 2159 :     low_risk low_risk       


row 3134 :     low_risk high_risk      incorrect
row 3135 :     low_risk high_risk      incorrect
row 3136 :     low_risk high_risk      incorrect
row 3137 :    high_risk high_risk      
row 3138 :     low_risk high_risk      incorrect
row 3139 :     low_risk high_risk      incorrect
row 3140 :    high_risk high_risk      
row 3141 :    high_risk high_risk      
row 3142 :    high_risk high_risk      
row 3143 :     low_risk high_risk      incorrect
row 3144 :    high_risk high_risk      
row 3145 :    high_risk high_risk      
row 3146 :     low_risk high_risk      incorrect
row 3147 :    high_risk high_risk      
row 3148 :     low_risk high_risk      incorrect
row 3149 :     low_risk high_risk      incorrect
row 3150 :    high_risk high_risk      
row 3151 :     low_risk high_risk      incorrect
row 3152 :     low_risk high_risk      incorrect
row 3153 :    high_risk high_risk      
row 3154 :     low_risk high_risk      incorrect
row 3155 :     low_risk high_risk      incorrect
row

row 3634 :     low_risk high_risk      incorrect
row 3635 :     low_risk high_risk      incorrect
row 3636 :    high_risk high_risk      
row 3637 :     low_risk high_risk      incorrect
row 3638 :     low_risk high_risk      incorrect
row 3639 :    high_risk high_risk      
row 3640 :     low_risk high_risk      incorrect
row 3641 :    high_risk high_risk      
row 3642 :     low_risk high_risk      incorrect
row 3643 :    high_risk high_risk      
row 3644 :    high_risk high_risk      
row 3645 :     low_risk high_risk      incorrect
row 3646 :     low_risk high_risk      incorrect
row 3647 :     low_risk high_risk      incorrect
row 3648 :     low_risk high_risk      incorrect
row 3649 :     low_risk high_risk      incorrect
row 3650 :     low_risk high_risk      incorrect
row 3651 :     low_risk high_risk      incorrect
row 3652 :    high_risk high_risk      
row 3653 :     low_risk high_risk      incorrect
row 3654 :    high_risk high_risk      
row 3655 :     low_risk high_risk 

row 4633 :     low_risk high_risk      incorrect
row 4634 :    high_risk high_risk      
row 4635 :     low_risk high_risk      incorrect
row 4636 :     low_risk high_risk      incorrect
row 4637 :    high_risk high_risk      
row 4638 :     low_risk high_risk      incorrect
row 4639 :     low_risk high_risk      incorrect
row 4640 :    high_risk high_risk      
row 4641 :     low_risk high_risk      incorrect
row 4642 :     low_risk high_risk      incorrect
row 4643 :     low_risk high_risk      incorrect
row 4644 :     low_risk high_risk      incorrect
row 4645 :     low_risk high_risk      incorrect
row 4646 :    high_risk high_risk      
row 4647 :    high_risk high_risk      
row 4648 :     low_risk high_risk      incorrect
row 4649 :     low_risk high_risk      incorrect
row 4650 :     low_risk high_risk      incorrect
row 4651 :     low_risk high_risk      incorrect
row 4652 :     low_risk high_risk      incorrect
row 4653 :     low_risk high_risk      incorrect
row 4654 :     l

2470

In [48]:
#Get the score of our Logistic Regression model predict
print(f"Testing Data Score: {classifier.score(X_test, y_test)}")

Testing Data Score: 0.5253083794130158


# Train a Random Forest Classifier model and print the model score

In [49]:
# Train a Random Forest Classifier model and print the model score
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier(random_state=1, n_estimators=500).fit(X_train, y_train)
print(f'Training Score: {clf.score(X_train, y_train)}')
print(f'Testing Score: {clf.score(X_test, y_test)}')

Training Score: 1.0
Testing Score: 0.6156954487452148


In [50]:
#Use our Random Forest Classifier model to test the 2020Q1loans accuracy
predictions_rf = clf.predict(X_test)
print(f"The new point was classified as: {predictions_rf}")

The new point was classified as: [0 0 0 ... 1 0 1]


In [51]:
#See the compare between predictions and Actual result
pd.DataFrame({"Prediction": predictions_rf, "Actual":y_test})

Prediction  Actual
0              0       0
1              0       0
2              0       0
3              0       0
4              0       0
...          ...     ...
4697           1       1
4698           1       1
4699           1       1
4700           0       1
4701           1       1

[4702 rows x 2 columns]

In [52]:
predicted_labels_rf = clf.predict(X_test)
actual_labels_rf = y_test
loan_risk = ['low_risk','high_risk']

def compare_labels(predicted_labels, actual_labels):
    """ a more neatly formatted comparison """
    NUM_LABELS = len(predicted_labels)
    num_correct = 0
    
    for i in range(NUM_LABELS):
        p = int(round(predicted_labels[i]))         # round protects from fp error 
        a = int(round(actual_labels[i]))
        result = "incorrect"
        if p == a:  # if they match,
            result = ""       # no longer incorrect
            num_correct += 1  # and we count a match!

        print(f"row {i:>3d} : {loan_risk[p]:>12s} {loan_risk[a]:<12s}   {result}")   

    print()
    print("Correct:", num_correct, "out of", NUM_LABELS)
    return num_correct

compare_labels(predicted_labels_rf,actual_labels_rf)

row   0 :     low_risk low_risk       
row   1 :     low_risk low_risk       
row   2 :     low_risk low_risk       
row   3 :     low_risk low_risk       
row   4 :     low_risk low_risk       
row   5 :     low_risk low_risk       
row   6 :     low_risk low_risk       
row   7 :     low_risk low_risk       
row   8 :     low_risk low_risk       
row   9 :     low_risk low_risk       
row  10 :     low_risk low_risk       
row  11 :     low_risk low_risk       
row  12 :     low_risk low_risk       
row  13 :     low_risk low_risk       
row  14 :    high_risk low_risk       incorrect
row  15 :     low_risk low_risk       
row  16 :     low_risk low_risk       
row  17 :     low_risk low_risk       
row  18 :     low_risk low_risk       
row  19 :     low_risk low_risk       
row  20 :     low_risk low_risk       
row  21 :     low_risk low_risk       
row  22 :     low_risk low_risk       
row  23 :     low_risk low_risk       
row  24 :     low_risk low_risk       
row  25 :     lo

row 1387 :    high_risk low_risk       incorrect
row 1388 :     low_risk low_risk       
row 1389 :     low_risk low_risk       
row 1390 :     low_risk low_risk       
row 1391 :     low_risk low_risk       
row 1392 :     low_risk low_risk       
row 1393 :     low_risk low_risk       
row 1394 :     low_risk low_risk       
row 1395 :     low_risk low_risk       
row 1396 :     low_risk low_risk       
row 1397 :     low_risk low_risk       
row 1398 :     low_risk low_risk       
row 1399 :     low_risk low_risk       
row 1400 :     low_risk low_risk       
row 1401 :     low_risk low_risk       
row 1402 :     low_risk low_risk       
row 1403 :     low_risk low_risk       
row 1404 :     low_risk low_risk       
row 1405 :     low_risk low_risk       
row 1406 :     low_risk low_risk       
row 1407 :     low_risk low_risk       
row 1408 :     low_risk low_risk       
row 1409 :     low_risk low_risk       
row 1410 :     low_risk low_risk       
row 1411 :    high_risk low_ris

row 2387 :    high_risk high_risk      
row 2388 :     low_risk high_risk      incorrect
row 2389 :    high_risk high_risk      
row 2390 :     low_risk high_risk      incorrect
row 2391 :    high_risk high_risk      
row 2392 :     low_risk high_risk      incorrect
row 2393 :     low_risk high_risk      incorrect
row 2394 :     low_risk high_risk      incorrect
row 2395 :     low_risk high_risk      incorrect
row 2396 :    high_risk high_risk      
row 2397 :    high_risk high_risk      
row 2398 :     low_risk high_risk      incorrect
row 2399 :    high_risk high_risk      
row 2400 :     low_risk high_risk      incorrect
row 2401 :     low_risk high_risk      incorrect
row 2402 :     low_risk high_risk      incorrect
row 2403 :     low_risk high_risk      incorrect
row 2404 :     low_risk high_risk      incorrect
row 2405 :     low_risk high_risk      incorrect
row 2406 :     low_risk high_risk      incorrect
row 2407 :    high_risk high_risk      
row 2408 :     low_risk high_risk 

row 3244 :    high_risk high_risk      
row 3245 :    high_risk high_risk      
row 3246 :     low_risk high_risk      incorrect
row 3247 :     low_risk high_risk      incorrect
row 3248 :    high_risk high_risk      
row 3249 :     low_risk high_risk      incorrect
row 3250 :     low_risk high_risk      incorrect
row 3251 :    high_risk high_risk      
row 3252 :     low_risk high_risk      incorrect
row 3253 :     low_risk high_risk      incorrect
row 3254 :    high_risk high_risk      
row 3255 :    high_risk high_risk      
row 3256 :     low_risk high_risk      incorrect
row 3257 :    high_risk high_risk      
row 3258 :     low_risk high_risk      incorrect
row 3259 :     low_risk high_risk      incorrect
row 3260 :     low_risk high_risk      incorrect
row 3261 :     low_risk high_risk      incorrect
row 3262 :     low_risk high_risk      incorrect
row 3263 :     low_risk high_risk      incorrect
row 3264 :     low_risk high_risk      incorrect
row 3265 :    high_risk high_risk 

row 3451 :     low_risk high_risk      incorrect
row 3452 :     low_risk high_risk      incorrect
row 3453 :     low_risk high_risk      incorrect
row 3454 :    high_risk high_risk      
row 3455 :    high_risk high_risk      
row 3456 :    high_risk high_risk      
row 3457 :    high_risk high_risk      
row 3458 :    high_risk high_risk      
row 3459 :     low_risk high_risk      incorrect
row 3460 :     low_risk high_risk      incorrect
row 3461 :     low_risk high_risk      incorrect
row 3462 :     low_risk high_risk      incorrect
row 3463 :    high_risk high_risk      
row 3464 :     low_risk high_risk      incorrect
row 3465 :     low_risk high_risk      incorrect
row 3466 :     low_risk high_risk      incorrect
row 3467 :     low_risk high_risk      incorrect
row 3468 :     low_risk high_risk      incorrect
row 3469 :     low_risk high_risk      incorrect
row 3470 :     low_risk high_risk      incorrect
row 3471 :    high_risk high_risk      
row 3472 :     low_risk high_risk 

row 4385 :     low_risk high_risk      incorrect
row 4386 :     low_risk high_risk      incorrect
row 4387 :    high_risk high_risk      
row 4388 :    high_risk high_risk      
row 4389 :    high_risk high_risk      
row 4390 :    high_risk high_risk      
row 4391 :     low_risk high_risk      incorrect
row 4392 :    high_risk high_risk      
row 4393 :    high_risk high_risk      
row 4394 :    high_risk high_risk      
row 4395 :    high_risk high_risk      
row 4396 :     low_risk high_risk      incorrect
row 4397 :    high_risk high_risk      
row 4398 :    high_risk high_risk      
row 4399 :    high_risk high_risk      
row 4400 :    high_risk high_risk      
row 4401 :    high_risk high_risk      
row 4402 :    high_risk high_risk      
row 4403 :     low_risk high_risk      incorrect
row 4404 :     low_risk high_risk      incorrect
row 4405 :     low_risk high_risk      incorrect
row 4406 :     low_risk high_risk      incorrect
row 4407 :     low_risk high_risk      incorrect

2895

In [53]:
#Get the score of our Random Forest Classifier model predict
print(f"Testing Data Score: {clf.score(X_test, y_test)}")

Testing Data Score: 0.6156954487452148


# Scale the data

In [54]:
# Scale the data
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler().fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [55]:
# Train the Logistic Regression model on the scaled data and print the model score

from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression()
classifier.fit(X_train_scaled, y_train)

/Users/chiachinghsieh/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:765: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


LogisticRegression()

In [56]:
print(f"Training Data Score: {classifier.score(X_train_scaled, y_train)}")
print(f"Testing Data Score: {classifier.score(X_test_scaled, y_test)}")

Training Data Score: 0.713136288998358
Testing Data Score: 0.560187154402382


In [57]:
predicted_labels = classifier.predict(X_test_scaled)
actual_labels = y_test
loan_risk = ['low_risk','high_risk']

def compare_labels(predicted_labels, actual_labels):
    """ a more neatly formatted comparison """
    NUM_LABELS = len(predicted_labels)
    num_correct = 0
    
    for i in range(NUM_LABELS):
        p = int(round(predicted_labels[i]))         # round protects from fp error 
        a = int(round(actual_labels[i]))
        result = "incorrect"
        if p == a:  # if they match,
            result = ""       # no longer incorrect
            num_correct += 1  # and we count a match!

        print(f"row {i:>3d} : {loan_risk[p]:>12s} {loan_risk[a]:<12s}   {result}")   

    print()
    print("Correct:", num_correct, "out of", NUM_LABELS)
    return num_correct

compare_labels(predicted_labels,actual_labels)

row   0 :    high_risk low_risk       incorrect
row   1 :     low_risk low_risk       
row   2 :     low_risk low_risk       
row   3 :    high_risk low_risk       incorrect
row   4 :    high_risk low_risk       incorrect
row   5 :    high_risk low_risk       incorrect
row   6 :    high_risk low_risk       incorrect
row   7 :    high_risk low_risk       incorrect
row   8 :    high_risk low_risk       incorrect
row   9 :     low_risk low_risk       
row  10 :    high_risk low_risk       incorrect
row  11 :    high_risk low_risk       incorrect
row  12 :    high_risk low_risk       incorrect
row  13 :    high_risk low_risk       incorrect
row  14 :    high_risk low_risk       incorrect
row  15 :    high_risk low_risk       incorrect
row  16 :    high_risk low_risk       incorrect
row  17 :     low_risk low_risk       
row  18 :    high_risk low_risk       incorrect
row  19 :    high_risk low_risk       incorrect
row  20 :    high_risk low_risk       incorrect
row  21 :    high_risk low_r

row 1140 :     low_risk low_risk       
row 1141 :    high_risk low_risk       incorrect
row 1142 :     low_risk low_risk       
row 1143 :    high_risk low_risk       incorrect
row 1144 :     low_risk low_risk       
row 1145 :    high_risk low_risk       incorrect
row 1146 :    high_risk low_risk       incorrect
row 1147 :    high_risk low_risk       incorrect
row 1148 :     low_risk low_risk       
row 1149 :    high_risk low_risk       incorrect
row 1150 :    high_risk low_risk       incorrect
row 1151 :    high_risk low_risk       incorrect
row 1152 :    high_risk low_risk       incorrect
row 1153 :    high_risk low_risk       incorrect
row 1154 :     low_risk low_risk       
row 1155 :    high_risk low_risk       incorrect
row 1156 :    high_risk low_risk       incorrect
row 1157 :    high_risk low_risk       incorrect
row 1158 :    high_risk low_risk       incorrect
row 1159 :    high_risk low_risk       incorrect
row 1160 :    high_risk low_risk       incorrect
row 1161 :    hi

row 2140 :    high_risk low_risk       incorrect
row 2141 :    high_risk low_risk       incorrect
row 2142 :    high_risk low_risk       incorrect
row 2143 :    high_risk low_risk       incorrect
row 2144 :    high_risk low_risk       incorrect
row 2145 :     low_risk low_risk       
row 2146 :    high_risk low_risk       incorrect
row 2147 :    high_risk low_risk       incorrect
row 2148 :    high_risk low_risk       incorrect
row 2149 :    high_risk low_risk       incorrect
row 2150 :    high_risk low_risk       incorrect
row 2151 :    high_risk low_risk       incorrect
row 2152 :    high_risk low_risk       incorrect
row 2153 :    high_risk low_risk       incorrect
row 2154 :    high_risk low_risk       incorrect
row 2155 :    high_risk low_risk       incorrect
row 2156 :    high_risk low_risk       incorrect
row 2157 :    high_risk low_risk       incorrect
row 2158 :    high_risk low_risk       incorrect
row 2159 :    high_risk low_risk       incorrect
row 2160 :    high_risk low_r

row 3139 :    high_risk high_risk      
row 3140 :    high_risk high_risk      
row 3141 :    high_risk high_risk      
row 3142 :    high_risk high_risk      
row 3143 :    high_risk high_risk      
row 3144 :    high_risk high_risk      
row 3145 :    high_risk high_risk      
row 3146 :    high_risk high_risk      
row 3147 :    high_risk high_risk      
row 3148 :    high_risk high_risk      
row 3149 :    high_risk high_risk      
row 3150 :    high_risk high_risk      
row 3151 :    high_risk high_risk      
row 3152 :    high_risk high_risk      
row 3153 :    high_risk high_risk      
row 3154 :    high_risk high_risk      
row 3155 :    high_risk high_risk      
row 3156 :    high_risk high_risk      
row 3157 :    high_risk high_risk      
row 3158 :    high_risk high_risk      
row 3159 :    high_risk high_risk      
row 3160 :    high_risk high_risk      
row 3161 :    high_risk high_risk      
row 3162 :    high_risk high_risk      
row 3163 :    high_risk high_risk      


row 4139 :    high_risk high_risk      
row 4140 :    high_risk high_risk      
row 4141 :    high_risk high_risk      
row 4142 :    high_risk high_risk      
row 4143 :    high_risk high_risk      
row 4144 :    high_risk high_risk      
row 4145 :    high_risk high_risk      
row 4146 :    high_risk high_risk      
row 4147 :    high_risk high_risk      
row 4148 :    high_risk high_risk      
row 4149 :    high_risk high_risk      
row 4150 :    high_risk high_risk      
row 4151 :    high_risk high_risk      
row 4152 :    high_risk high_risk      
row 4153 :    high_risk high_risk      
row 4154 :    high_risk high_risk      
row 4155 :    high_risk high_risk      
row 4156 :    high_risk high_risk      
row 4157 :    high_risk high_risk      
row 4158 :    high_risk high_risk      
row 4159 :    high_risk high_risk      
row 4160 :    high_risk high_risk      
row 4161 :    high_risk high_risk      
row 4162 :    high_risk high_risk      
row 4163 :    high_risk high_risk      


2634

In [58]:
print(f"Testing Data Score: {classifier.score(X_test_scaled, y_test)}")

Testing Data Score: 0.560187154402382


In [59]:
# Train a Random Forest Classifier model on the scaled data and print the model score

from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier(random_state=1, n_estimators=500).fit(X_train_scaled, y_train)
print(f'Training Score: {clf.score(X_train_scaled, y_train)}')
print(f'Testing Score: {clf.score(X_test_scaled, y_test)}')

Training Score: 1.0
Testing Score: 0.6163334751169715


In [60]:
print(f"Training Data Score: {clf.score(X_train_scaled, y_train)}")
print(f"Testing Data Score: {clf.score(X_test_scaled, y_test)}")

Training Data Score: 1.0
Testing Data Score: 0.6163334751169715


In [61]:
predicted_labels_rf = clf.predict(X_test_scaled)
actual_labels_rf = y_test
loan_risk = ['low_risk','high_risk']

def compare_labels(predicted_labels, actual_labels):
    """ a more neatly formatted comparison """
    NUM_LABELS = len(predicted_labels)
    num_correct = 0
    
    for i in range(NUM_LABELS):
        p = int(round(predicted_labels[i]))         # round protects from fp error 
        a = int(round(actual_labels[i]))
        result = "incorrect"
        if p == a:  # if they match,
            result = ""       # no longer incorrect
            num_correct += 1  # and we count a match!

        print(f"row {i:>3d} : {loan_risk[p]:>12s} {loan_risk[a]:<12s}   {result}")   

    print()
    print("Correct:", num_correct, "out of", NUM_LABELS)
    return num_correct

compare_labels(predicted_labels_rf,actual_labels_rf)

row   0 :     low_risk low_risk       
row   1 :     low_risk low_risk       
row   2 :     low_risk low_risk       
row   3 :     low_risk low_risk       
row   4 :     low_risk low_risk       
row   5 :     low_risk low_risk       
row   6 :     low_risk low_risk       
row   7 :     low_risk low_risk       
row   8 :     low_risk low_risk       
row   9 :     low_risk low_risk       
row  10 :     low_risk low_risk       
row  11 :     low_risk low_risk       
row  12 :     low_risk low_risk       
row  13 :     low_risk low_risk       
row  14 :    high_risk low_risk       incorrect
row  15 :     low_risk low_risk       
row  16 :     low_risk low_risk       
row  17 :     low_risk low_risk       
row  18 :     low_risk low_risk       
row  19 :     low_risk low_risk       
row  20 :     low_risk low_risk       
row  21 :     low_risk low_risk       
row  22 :     low_risk low_risk       
row  23 :     low_risk low_risk       
row  24 :     low_risk low_risk       
row  25 :     lo

row 1399 :     low_risk low_risk       
row 1400 :     low_risk low_risk       
row 1401 :     low_risk low_risk       
row 1402 :     low_risk low_risk       
row 1403 :     low_risk low_risk       
row 1404 :     low_risk low_risk       
row 1405 :     low_risk low_risk       
row 1406 :     low_risk low_risk       
row 1407 :     low_risk low_risk       
row 1408 :     low_risk low_risk       
row 1409 :     low_risk low_risk       
row 1410 :     low_risk low_risk       
row 1411 :    high_risk low_risk       incorrect
row 1412 :     low_risk low_risk       
row 1413 :     low_risk low_risk       
row 1414 :     low_risk low_risk       
row 1415 :     low_risk low_risk       
row 1416 :    high_risk low_risk       incorrect
row 1417 :     low_risk low_risk       
row 1418 :    high_risk low_risk       incorrect
row 1419 :     low_risk low_risk       
row 1420 :     low_risk low_risk       
row 1421 :     low_risk low_risk       
row 1422 :     low_risk low_risk       
row 1423 :   

row 2399 :    high_risk high_risk      
row 2400 :     low_risk high_risk      incorrect
row 2401 :     low_risk high_risk      incorrect
row 2402 :     low_risk high_risk      incorrect
row 2403 :     low_risk high_risk      incorrect
row 2404 :     low_risk high_risk      incorrect
row 2405 :     low_risk high_risk      incorrect
row 2406 :     low_risk high_risk      incorrect
row 2407 :    high_risk high_risk      
row 2408 :     low_risk high_risk      incorrect
row 2409 :     low_risk high_risk      incorrect
row 2410 :     low_risk high_risk      incorrect
row 2411 :     low_risk high_risk      incorrect
row 2412 :    high_risk high_risk      
row 2413 :    high_risk high_risk      
row 2414 :     low_risk high_risk      incorrect
row 2415 :    high_risk high_risk      
row 2416 :    high_risk high_risk      
row 2417 :     low_risk high_risk      incorrect
row 2418 :     low_risk high_risk      incorrect
row 2419 :     low_risk high_risk      incorrect
row 2420 :     low_risk h

row 3398 :    high_risk high_risk      
row 3399 :    high_risk high_risk      
row 3400 :     low_risk high_risk      incorrect
row 3401 :    high_risk high_risk      
row 3402 :     low_risk high_risk      incorrect
row 3403 :    high_risk high_risk      
row 3404 :     low_risk high_risk      incorrect
row 3405 :     low_risk high_risk      incorrect
row 3406 :     low_risk high_risk      incorrect
row 3407 :     low_risk high_risk      incorrect
row 3408 :     low_risk high_risk      incorrect
row 3409 :    high_risk high_risk      
row 3410 :     low_risk high_risk      incorrect
row 3411 :    high_risk high_risk      
row 3412 :     low_risk high_risk      incorrect
row 3413 :     low_risk high_risk      incorrect
row 3414 :     low_risk high_risk      incorrect
row 3415 :     low_risk high_risk      incorrect
row 3416 :     low_risk high_risk      incorrect
row 3417 :     low_risk high_risk      incorrect
row 3418 :     low_risk high_risk      incorrect
row 3419 :     low_risk h

2898

In [62]:
print(f"Testing Data Score: {clf.score(X_test_scaled, y_test)}")

Testing Data Score: 0.6163334751169715


## Conculstion

## Use whole 2019 and 2021 dataset to train the model
# Before Scale the data
The result 
@Logistic Regression model : get Testing Score:0.525 and test result is Correct: 2470 out of 4702
        
        
@Random Forest Classifier : get Testing Score: 0.61569 and test result is  Correct: 2895 out of 4702 
 
 
we can said that before scale the data, Random Forest model have better perform than Logistic Regressio model.

# After  Scale the data

The result 
@Logistic Regression model : get Testing Score: 0.560 and test result is  Correct: 2634 out of 4702 
        
@Random Forest Classifier : get Testing Score: 0.616 and test result is  Correct: 2898 out of 4702
            
we can see that after scale the data, the accuracy have been improved but not so much.

# Maybe the next setp we select feature and try the K-Mean model on this dataset.


